File: hyperbolic/example/prep_grid/model1.m
clc;
clear all;
close all;

% p1        p2        p5            p6
% ...........         ...............
%           .         .
%           .         .
%           ...........    
%           p3        p4

flag_printf = 1;

nx = 801;

p1 = 1;
p2 = 201;
p3 = 301;
p4 = 501;
p5 = 601;
p6 = 801;

dh = 1;
origin_x = 0;
origin_z = 0;

bz = zeros(nx,2);
% NOTE: x direction tight
% so coord x incremental with x
for i=p1:p2
    bz(i,1) = origin_x + (i-1)*dh;
    bz(i,2) = origin_z;
end
for i=p2+1:p3
    bz(i,1) = origin_x + bz(p2,1) + cos(0.5*pi*70/90)*(i-p2)*dh;
    bz(i,2) = origin_z + bz(p2,2) - sin(0.5*pi*70/90)*(i-p2)*dh;
end
for i=p3+1:p4
    bz(i,1) = origin_x + bz(p3,1) + (i-p3)*dh;
    bz(i,2) = origin_z + bz(p3,2);
end
for i=p4+1:p5
    bz(i,1) = origin_x + bz(p4,1) + cos(0.5*pi*70/90)*(i-p4)*dh;
    bz(i,2) = origin_z + bz(p4,2) + sin(0.5*pi*70/90)*(i-p4)*dh;
end
for i=p5+1:p6
    bz(i,1) = origin_x + bz(p5,1) + (i-p5)*dh;
    bz(i,2) = origin_z + bz(p5,2);
end
% bz(:,2) = smooth(bz(:,2),20);
% bz(:,1) = smooth(bz(:,1),20);
if flag_printf
    figure(1)   
    plot(bz(:,1),bz(:,2),'k');
    hold on;

    plot(100.0,   0.0  , 'kv');
    plot(200.0,   0.0  , 'kv');
    plot(234.202,  -93.9693, 'kv');
    plot(334.202,  -93.9693, 'kv');
    plot(434.202,  -93.9693,'kv');
    plot(468.404,   0    , 'kv');
    plot(568.404,   0    , 'kv');
    text(100.0+10,  0.0  +20,'R1',FontSize=15);
    text(200.0+10,  0.0  +20,'R2',FontSize=15);
    text(234.202+10,-93.9693+20,'R3',FontSize=15);
    text(334.202+10,-93.9693+20,'R4',FontSize=15);
    text(434.202+10,-93.9693+20,'R5',FontSize=15);
    text(468.404+10, 0    +20,'R6',FontSize=15);
    text(568.404+10, 0    +20,'R7',FontSize=15);
    plot(150,-50,'k*');
    axis equal;
    set(gcf,'Position',[200,200,650,400]);
    xlabel('X axis (m)',FontSize=15,FontWeight='bold');
    ylabel('Y axis (m)',FontSize=15,FontWeight='bold');
    xlim([-100,800]);
    ylim([-400,100]);
    %title('concave step model',FontSize=15);
    set(gca,'layer','top');
    set(gca,'FontSize',10,FontWeight='bold');
    set(gcf,'color','white','renderer','painters');
    text(200,-150,'Vp=3.0 km s^{-1}',FontSize=15);
    text(200,-200,'Vs=2.0 km s^{-1}',FontSize=15);
    text(200,-250,'\rho=1.5 g cm^{-3}',FontSize=15);
end

% creat data file
file_name = '../data_file_2d.txt';
fid=fopen(file_name,'w'); % Output file name 
fprintf(fid,'# nx number\n'); 
fprintf(fid,'%d\n',nx);
fprintf(fid,'# bz coords\n'); 
for i=1:nx
  fprintf(fid,'%.9e %.9e\n',bz(i,1),bz(i,2));
end

if flag_printf
   print(gcf,'model1.png','-r400','-dpng');
 end

File: hyperbolic/example/prep_grid/model1.py
import sys, os
import numpy as np
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'plotting'))
from export_c import export_bdry_1_bz
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

def main():
    flag_printf = 1

    nx = 401
    ny = 301

    p1 = 0      # 1 in MATLAB, 0-based here
    p2 = 100     # 101 in MATLAB
    p3 = 150     # 151 in MATLAB
    p4 = 250     # 251 in MATLAB
    p5 = 300     # 301 in MATLAB
    p6 = 400     # 401 in MATLAB

    dh = 10.0
    origin_x = 0.0
    origin_y = 0.0
    origin_z = 0.0

    # bz shape is (ny, nx, 3) as per export_bdry_1_bz's expectations
    bz = np.zeros((ny, nx, 3))

    for j in range(ny):
        # First segment: i from p1 to p2 inclusive
        for i in range(p1, p2 + 1):
            bz[j, i, 0] = origin_x + i * dh
            bz[j, i, 1] = origin_y + j * dh
            bz[j, i, 2] = origin_z

        # Second segment: i from p2+1 to p3 inclusive
        for i in range(p2 + 1, p3 + 1):
            bz[j, i, 0] = origin_x + bz[j, p2, 0] + np.cos(0.5 * np.pi * 90 / 90) * (i - p2) * dh
            bz[j, i, 1] = origin_y + j * dh
            bz[j, i, 2] = origin_z + bz[j, p2, 2] - np.sin(0.5 * np.pi * 90 / 90) * (i - p2) * dh

        # Third segment: i from p3+1 to p4 inclusive
        for i in range(p3 + 1, p4 + 1):
            bz[j, i, 0] = origin_x + bz[j, p3, 0] + (i - p3) * dh
            bz[j, i, 1] = origin_y + j * dh
            bz[j, i, 2] = origin_z + bz[j, p3, 2]

        # Fourth segment: i from p4+1 to p5 inclusive
        for i in range(p4 + 1, p5 + 1):
            bz[j, i, 0] = origin_x + bz[j, p4, 0] + np.cos(0.5 * np.pi * 90 / 90) * (i - p4) * dh
            bz[j, i, 1] = origin_y + j * dh
            bz[j, i, 2] = origin_z + bz[j, p4, 2] + np.sin(0.5 * np.pi * 90 / 90) * (i - p4) * dh

        # Fifth segment: i from p5+1 to p6 inclusive
        for i in range(p5 + 1, p6 + 1):
            bz[j, i, 0] = origin_x + bz[j, p5, 0] + (i - p5) * dh
            bz[j, i, 1] = origin_y + j * dh
            bz[j, i, 2] = origin_z + bz[j, p5, 2]

    file_name = '../data_file_3d.txt'
    export_bdry_1_bz(bz, nx, ny, file_name)

    if flag_printf:
        os.makedirs('../plot', exist_ok=True)
        stride = max(1, max(nx, ny) // 80)
        fig = plt.figure(figsize=(10, 6))
        ax = fig.add_subplot(111, projection='3d')
        X = bz[:, :, 0] / 1e3; Y = bz[:, :, 1] / 1e3; Z = bz[:, :, 2] / 1e3
        ax.plot_surface(X[::stride, ::stride], Y[::stride, ::stride],
                        Z[::stride, ::stride], cmap='jet', alpha=0.9, edgecolor='none')
        # MATLAB: plot3(bz(:,151,1)/1e3,...) — bz shape (ny,nx,3), mid at j=151
        mid_j = 151
        ax.plot(X[:, mid_j], Y[:, mid_j], Z[:, mid_j], 'r', linewidth=3)
        ax.set_xlabel('X axis (km)', fontweight='bold')
        ax.set_ylabel('Y axis (km)', fontweight='bold')
        ax.set_zlabel('Z axis (km)', fontweight='bold')
        ax.view_init(elev=30, azim=20)
        ax.set_box_aspect([1, 1, 0.5])
        ax.grid(False)
        mappable = plt.cm.ScalarMappable(cmap='jet')
        mappable.set_array(Z[::stride, ::stride].ravel())
        mappable.set_clim(Z.min(), Z.max())
        cb = fig.colorbar(mappable, ax=ax, shrink=0.6, pad=0.1)
        cb.set_label('(km)', fontweight='bold')
        fig.patch.set_facecolor('white')
        plt.tight_layout()
        plt.savefig('../plot/hyper_model1.png', dpi=300, bbox_inches='tight', facecolor='white')
        plt.close(fig)
        print('Saved: ../plot/hyper_model1.png')

if __name__ == "__main__":
    main()
